Source/WebCore/Modules/WebGPU/GPUDevice.cpp

```diff
@@ -59,27 +59,35 @@
 #include "GPUTextureDescriptor.h"
 #include "JSDOMPromiseDeferred.h"
 #include "JSGPUComputePipeline.h"
+#include "JSGPUDeviceLostInfo.h"
 #include "JSGPUInternalError.h"
 #include "JSGPUOutOfMemoryError.h"
 #include "JSGPUPipelineError.h"
 #include "JSGPURenderPipeline.h"
 #include "JSGPUValidationError.h"
+#include <pal/graphics/WebGPU/WebGPUAdapter.h>
 #include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(GPUDevice);
 
-GPUDevice::GPUDevice(ScriptExecutionContext* scriptExecutionContext, Ref<PAL::WebGPU::Device>&& backing)
+GPUDevice::GPUDevice(ScriptExecutionContext* scriptExecutionContext, Ref<PAL::WebGPU::Device>&& backing, Ref<PAL::WebGPU::Adapter>& adapter)
     : ActiveDOMObject { scriptExecutionContext }
     , m_lostPromise(makeUniqueRef<LostPromise>())
     , m_backing(WTFMove(backing))
     , m_queue(GPUQueue::create(Ref { m_backing->queue() }))
     , m_autoPipelineLayout(createPipelineLayout({ { "autoLayout"_s, }, { } }))
+    , m_adapter(adapter)
 {
+    if (!adapter->isValid())
+        makeInvalid();
 }
 
-GPUDevice::~GPUDevice() = default;
+GPUDevice::~GPUDevice()
+{
+    destroy();
```

Comment:
Are we sure we want this to resolve a promise? I don't think we do.

Reply:
With the introduction of `destruct` which is called from the destructor, I will only have destroy resolve the promise.